### PR TITLE
[NFC][TableGen] Elminate use of isalpha/isdigit from TGLexer

### PR DESCRIPTION
**Pull request by jurahul**
- Replace use of std::isalpha, std::isdigit, std:isxdigit with LLVM's StringExtras versions, to avoid possibly locale dependent behavior (e.g. glibc).
- Create helper function for common checks for valid identifier characters.

### CODE REVIEW

File: llvm/lib/TableGen/TGLexer.cpp [hunk 1]
```diff
@@ -38,6 +38,22 @@ struct PreprocessorDir {
...
+/// `First`, returns true if `C` is a valid first character of an identifier,
+/// else returns true if `C` is a valid non-first character of an identifier.
+/// Identifiers match the following regular expression:
+///   [a-zA-Z_][0-9a-zA-Z_]*
+enum IDCharPosition {
+  First = 0,
+  NotFirst = 1,
+};
+
+static bool IsValidIDChar(char C, IDCharPosition Position) {
```
Comment (MaskRay) [inline llvm/lib/TableGen/TGLexer.cpp:51]: `functionName` when adding new functions. Some functions don't stick with the convention. Just ignore them.

File: llvm/lib/TableGen/TGLexer.cpp [hunk 2]
```diff
@@ -51,14 +67,14 @@ static const char *lexMacroName(StringRef Str) {
 
   // Macro names start with [a-zA-Z_].
   const char *Next = Str.begin();
-  if (*Next != '_' && !isalpha(*Next))
+  if (!IsValidIDChar(*Next, First))
```
Comment (MaskRay) [inline llvm/lib/TableGen/TGLexer.cpp:70]: Just use a bool.

First/NotFirst that pollute the namespace isn't better than a boolean here.

**Review by MaskRay — APPROVED** [review]

File: llvm/lib/TableGen/TGLexer.cpp (hunk 1 above)
Comment (jurahul) [reply llvm/lib/TableGen/TGLexer.cpp]: Done.

File: llvm/lib/TableGen/TGLexer.cpp (hunk 2 above)
Comment (jurahul) [reply llvm/lib/TableGen/TGLexer.cpp]: Done.

**Pull request merged**